service/src/main/java/com/commercetools/pspadapter/payone/PaymentHandler.java

```diff
@@ -0,0 +1,111 @@
+package com.commercetools.pspadapter.payone;
+
+import com.commercetools.pspadapter.payone.domain.ctp.CommercetoolsQueryExecutor;
+import com.commercetools.pspadapter.payone.domain.ctp.PaymentWithCartLike;
+import com.commercetools.pspadapter.payone.domain.ctp.exceptions.NoCartLikeFoundException;
+import io.sphere.sdk.client.ErrorResponseException;
+import io.sphere.sdk.client.NotFoundException;
+import io.sphere.sdk.http.HttpStatusCode;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import java.util.ConcurrentModificationException;
+import java.util.concurrent.CompletionException;
+
+import static com.commercetools.pspadapter.tenant.TenantLoggerUtil.createLoggerName;
+import static io.sphere.sdk.http.HttpStatusCode.INTERNAL_SERVER_ERROR_500;
+import static java.lang.String.format;
+import static java.util.Optional.ofNullable;
+
+public class PaymentHandler {
+
+    /**
+     * How many times to retry if {@link ConcurrentModificationException} happens.
+     */
+    private static final int RETRIES_LIMIT = 20;
+    private static final int RETRY_DELAY = 100; // msec
+
+    private final String payoneInterfaceName;
+
+    private final CommercetoolsQueryExecutor commercetoolsQueryExecutor;
+    private final PaymentDispatcher paymentDispatcher;
+
+    private final Logger logger;
+
+    public PaymentHandler(String payoneInterfaceName, String tenantName,
+                          CommercetoolsQueryExecutor commercetoolsQueryExecutor, PaymentDispatcher paymentDispatcher) {
+        this.payoneInterfaceName = payoneInterfaceName;
+
+        this.commercetoolsQueryExecutor = commercetoolsQueryExecutor;
+        this.paymentDispatcher = paymentDispatcher;
+
+        this.logger = LoggerFactory.getLogger(createLoggerName(this.getClass(), tenantName));
+    }
+
+    /**
+     * Tries to handle the payment with the provided ID.
+     *
+     * @param paymentId identifies the payment to be processed
+     * @return the result of handling the payment
+     */
+    public PaymentHandleResult handlePayment(@Nonnull final String paymentId) {
+        try {
+            for (int i = 0; i < RETRIES_LIMIT; i++) {
+                try {
+                    final PaymentWithCartLike payment = commercetoolsQueryExecutor.getPaymentWithCartLike(paymentId);
+                    String paymentInterface = payment.getPayment().getPaymentMethodInfo().getPaymentInterface();
```

Comment:
[`CommercetoolsQueryExecutor#getPaymentWithCartLike(java.lang.String)`](https://github.com/commercetools/commercetools-payone-integration/blob/2c71659877198b14d45e08b7996303c7711bf851/service/src/main/java/com/commercetools/pspadapter/payone/domain/ctp/CommercetoolsQueryExecutor.java#L38-L38) returns a new instance or interrupted by exception.

Although i've renamed the instance name to be more reflective: 0a1ca55e